### ISSUE TITLE
tf.nn.softmax_cross_entropy_with_logits_v2 changes tensor shape

### ISSUE THREAD

**Issue by felixnext** [tensorflow/tensorflow#24425]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 16.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: No
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 1.12.0
- Python version: 3.5
- CUDA/cuDNN version: 9.0 / 7.4.1
- GPU model and memory: 1070 Ti / 8GB

I have a simple network that outputs logits via a dense layer in the model_fn (using estimator API):
```
logits = tf.layers.dense(inputs=dense1, units=2, activation=tf.nn.relu, name='logits')
```

I then calculate the loss for the training with:
```
if mode in (tf.estimator.ModeKeys.TRAIN, tf.estimator.ModeKeys.EVAL):
  print(logits.shape)
  print_op = tf.print("LABELS: ", logits, tf.shape(logits))
  with tf.control_dependencies([print_op]):
    logits = tf.identity(logits)
  loss = tf.nn.softmax_cross_entropy_with_logits_v2(labels=labels, logits=logits)
  tf.summary.scalar('cross_entropy', loss)
```

Which results in an error, as the tensor shapes appear not to match:
```bash
Caused by op 'Reshape_3', defined at:
  File "train.py", line 40, in <module>
    fire.Fire(train)
  File "/usr/local/lib/python3.5/dist-packages/fire/core.py", line 127, in Fire
    component_trace = _Fire(component, args, context, name)
  File "/usr/local/lib/python3.5/dist-packages/fire/core.py", line 366, in _Fire
    component, remaining_args)
  File "/usr/local/lib/python3.5/dist-packages/fire/core.py", line 542, in _CallCallable
    result = fn(*varargs, **kwargs)
  File "train.py", line 36, in train
    learning.train(model, input, model_name, exp_name, config_file, exp_restore)
  File "/data/code/bp-labeling/learning/learning/train.py", line 93, in train
    tf.estimator.train_and_evaluate(estimator, train_spec, eval_spec)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/estimator/training.py", line 471, in train_and_evaluate
    return executor.run()
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/estimator/training.py", line 610, in run
    return self.run_local()
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/estimator/training.py", line 711, in run_local
    saving_listeners=saving_listeners)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/estimator/estimator.py", line 354, in train
    loss = self._train_model(input_fn, hooks, saving_listeners)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/estimator/estimator.py", line 1207, in _train_model
    return self._train_model_default(input_fn, hooks, saving_listeners)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/estimator/estimator.py", line 1237, in _train_model_default
    features, labels, model_fn_lib.ModeKeys.TRAIN, self.config)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/estimator/estimator.py", line 1195, in _call_model_fn
    model_fn_results = self._model_fn(features=features, **kwargs)
  File "/data/code/bp-labeling/learning/learning/model.py", line 86, in model_fn
    return self.classifier_fn(features, images, labels, mode, logits, params, cls_count, global_step)
  File "/data/code/bp-labeling/learning/models/sample_net.py", line 83, in classifier_fn
    return tf.estimator.EstimatorSpec(mode, loss=loss, train_op=train_op, training_hooks=[hook])
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/estimator/model_fn.py", line 201, in __new__
    loss = array_ops.reshape(loss, [])
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/ops/gen_array_ops.py", line 6482, in reshape
    "Reshape", tensor=tensor, shape=shape, name=name)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/framework/op_def_library.py", line 787, in _apply_op_helper
    op_def=op_def)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/util/deprecation.py", line 488, in new_func
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/framework/ops.py", line 3274, in create_op
    op_def=op_def)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/framework/ops.py", line 1770, in __init__
    self._traceback = tf_stack.extract_stack()

InvalidArgumentError (see above for traceback): Input to reshape is a tensor with 10 values, but the requested shape has 1
	 [[node Reshape_3 (defined at /data/code/bp-labeling/learning/models/sample_net.py:83)  = Reshape[T=DT_FLOAT, Tshape=DT_INT32, _device="/job:localhost/replica:0/task:0/device:GPU:0"](softmax_cross_entropy_with_logits/Reshape_2, Reshape_3/shape)]]
	 [[{{node Reshape_3/_397}} = _Recv[client_terminated=false, recv_device="/job:localhost/replica:0/task:0/device:CPU:0", send_device="/job:localhost/replica:0/task:0/device:GPU:0", send_device_incarnation=1, tensor_name="edge_124_Reshape_3", tensor_type=DT_FLOAT, _device="/job:localhost/replica:0/task:0/device:CPU:0"]()]]
```

The `print` function is only called once with the following output:
```bash
LABELS:  [[0.255024195 0.00465549901]
 [0.550343394 0]
 [0.0432553887 0]
 ...
 [0.0975153148 0]
 [0 0]
 [0 0]] [1 2]
```
(output from `print(logits.shape)` is `(?, 2)`).

Which appears strange, as the shape is given as `[1, 2]`, which is obviously not the case.

If I change the loss function to:
```
loss = tf.losses.softmax_cross_entropy(onehot_labels=labels, logits=logits)
```

the code runs fine and the output of the print command is e.g.:
```bash
LABELS:  [[0 0]
 [0 20.4618092]
 [0 0]
 ...
 [0 1.23361325]
 [0 0]
 [0 0]] [10 2]
```

Am I missing something or is this a bug?

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by msymp** [comment]
Hello @felixnext , Can you please study this Stack Overflow post that exactly addresses your query:
https://stackoverflow.com/questions/41412335/tf-nn-softmax-cross-entropy-with-logits-error-logits-and-labels-must-be-same
Please let us know if you are satisfied the various solutions provided in the post. Thanks.

**Comment by msymp** [comment]
Hi Alfred,
Thank you for flagging this issue. It happened to one of initial
'reverse-shadow' exercises that I had initiated, and forgot to update the
response.
I have dispatched a Stack Overflow link just now to the author, that
addresses the query, and am awaiting their feedback. I shall close it upon
confirmation.
Thanks,
@msymp

**Comment by msymp** [comment]
Automatically closing due to lack of recent activity. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!
